### PR TITLE
[TTAHUB-835-809] finalize read only states for goals and objectives

### PR DESCRIPTION
**Pull request by thewatermethod**
## Description of change

The RTR has a lot of conditional UI, which can depend on a few different things. This PR made a few changes and bug fixes so that the UI matches the mockup.

## How to test

Proceed through [each screen in the mockup](https://app.mural.co/t/adhoccorporateworkspace2583/m/adhoccorporateworkspace2583/1650563114436/544b84e6ff0548ddd1f4dbeafff69f56e1ab50e2?sender=u775286a3a84f5c5204174211). Compare each screen with the UI in the RTR. (Sorry, it's a bit involved)

Note: [there are some bugs remaining](https://ocio-jira.acf.hhs.gov/browse/TTAHUB-1015), unrelated to this PR merged into main-ar-redesign has temporarily nerfed Activity Reports. You may have to work around them by directly modifying your DB, specifically the 1st, 2nd, and 4th bullets.

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-835
* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-809

### Every PR

- [ ] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: frontend/src/components/GoalForm/ObjectiveStatus.js [hunk 1]
```diff
@@ -17,10 +16,9 @@ export default function ObjectiveStatus({
     return null;
   }
 
-  // if the objective has been completed or is "in progress"
-  // we need a control to change status
-
-  const showDropdown = !(status.toLowerCase() === 'not started' && isOnReport);
+  // if the objective is "in progress" we need a control to change status
+  // or if the goal is 'closed'
+  const showDropdown = status === 'In Progress' && goalStatus !== 'Closed';
```
Comment (AdamAdHocTeam) [inline frontend/src/components/GoalForm/ObjectiveStatus.js:21]: Should this be goalStatus === 'Closed'? Im just wondering based on the above comment.

File: frontend/src/components/GoalForm/ObjectiveTopics.js [hunk 2]
```diff
@@ -15,27 +15,30 @@ export default function ObjectiveTopics({
...
+  goalStatus,
   inputName,
   isOnReport,
   isOnApprovedReport,
   isLoading,
 }) {
   const initialSelection = useRef(topics.length);
 
-  const readOnly = status === 'Suspended' || (status === 'Not Started' && isOnReport);
+  const readOnly = useMemo(() => status === 'Suspended' || (status === 'Not Started' && isOnReport) || (status === 'Completed' && goalStatus === 'Closed'), [goalStatus, isOnReport, status]);
```
Comment (AdamAdHocTeam) [inline frontend/src/components/GoalForm/ObjectiveTopics.js:26]: Above you have '(status === 'Not Started' && isOnReport)' AND ' (status === 'In Progress' && isOnReport)'.

Do you need to check 'In Progress' with isOnReport (or I might be confused).

File: frontend/src/components/GoalForm/ResourceRepeater.js [hunk 3]
```diff
@@ -19,10 +20,11 @@ export default function ResourceRepeater({
   status,
   isOnReport,
   isLoading,
+  goalStatus,
 }) {
   const resourcesWrapper = useRef();
 
-  const readOnly = status === 'Suspended' || (status === 'Not Started' && isOnReport);
+  const readOnly = status === 'Suspended' || (status === 'Not Started' && isOnReport) || (status === 'Completed' && goalStatus === 'Closed');
```
Comment (AdamAdHocTeam) [inline frontend/src/components/GoalForm/ResourceRepeater.js:27]: Again do we care about 'In Progress' here?

File: frontend/src/components/GoalForm/SpecialistRole.js [hunk 4]
```diff
@@ -1,95 +1,133 @@
...
   const initialSelectedRolesLength = useRef(selectedRoles.length);
 
-  // if there is only one option, we just set the objectives to be
-  // that value without any UI
-  if (options.length === 1) {
-    return null;
-  }
-
-  const readOnly = isOnApprovedReport || status === 'Suspended' || (status === 'Not Started' && isOnReport);
+  const readOnly = useMemo(() => status === 'Suspended' || (status === 'Not Started' && isOnReport) || (status === 'Completed' && goalStatus === 'Closed'), [goalStatus, isOnReport, status]);
```
Comment (AdamAdHocTeam) [inline frontend/src/components/GoalForm/SpecialistRole.js:25]: Same q here as above.

File: frontend/src/pages/ActivityReport/Pages/components/RecipientReviewSection.js [hunk 5]
```diff
@@ -28,14 +28,14 @@ const RecipientReviewSection = () => {
         const objectives = goal.objectives || [];
         return (
           <div key={`review-${goal.id}`}>
-            <div className="grid-row margin-bottom-3 desktop:margin-bottom-0 margin-top-2">
-              <span>
+            <div className="margin-bottom-3 desktop:margin-bottom-0 margin-top-2">
+              <div>
```
Comment (AdamAdHocTeam) [inline frontend/src/pages/ActivityReport/Pages/components/RecipientReviewSection.js:32]: Why <div> over <span>?

File: src/services/goals.js [hunk 6]
```diff
@@ -359,9 +365,8 @@ export function reduceObjectives(newObjectives, currentObjectives = []) {
       ids: [id],
       // Make sure we pass back a list of recipient ids for subsequent saves.
       recipientIds: [objective.getDataValue('otherEntityId')],
-      ttaProvided,
+      // ttaProvided
```
Comment (AdamAdHocTeam) [inline src/services/goals.js:368]: Should this be commented out?

**Review by AdamAdHocTeam — COMMENTED** [review]
Not sure if this is fixed on another PR:

1. Create a new goal with a single objective and attachment.
2. Go back to the RTR and edit the objective.
3. Try to add another attachment.

You should get a 400, it looks like we aren't passing back the objectives ids from the network tab.

**Comment by thewatermethod** [comment]
> Not sure if this is fixed on another PR:

>    Create a new goal with a single objective and attachment.
>    Go back to the RTR and edit the objective.
>    Try to add another attachment.

> You should get a 400, it looks like we aren't passing back the objectives ids from the network tab.

yeah, this is unrelated to this PR but I'm fixing it on #1036

File: frontend/src/pages/ActivityReport/Pages/components/RecipientReviewSection.js (hunk 5 above)
Comment (thewatermethod) [reply frontend/src/pages/ActivityReport/Pages/components/RecipientReviewSection.js:32]: Wanted those natural block-level element stylings

File: frontend/src/components/GoalForm/ObjectiveTopics.js (hunk 2 above)
Comment (thewatermethod) [reply frontend/src/components/GoalForm/ObjectiveTopics.js:26]: This logic refects the mockup

File: frontend/src/components/GoalForm/ResourceRepeater.js (hunk 3 above)
Comment (thewatermethod) [reply frontend/src/components/GoalForm/ResourceRepeater.js:27]: This logic refects the mockup

File: frontend/src/components/GoalForm/SpecialistRole.js (hunk 4 above)
Comment (thewatermethod) [reply frontend/src/components/GoalForm/SpecialistRole.js:25]: This logic refects the mockup

File: src/services/goals.js (hunk 6 above)
Comment (thewatermethod) [reply src/services/goals.js]: I removed the commented out line (this method is called in places where there will be no TTA provided or we don't care about that)

**Review by AdamAdHocTeam — COMMENTED** [review]
When viewing a objective on a approved AR should the resource links be URL's?

image

In mockup:

image

**Review by AdamAdHocTeam — COMMENTED** [review]
Used one objective attachment of two on an approved report but see this on the objective edit screen.

image

Can you let me know if you can reproduce. I had to manually change DB data to get it to work so it might just be me.

**Comment by thewatermethod** [comment]
> Used one objective attachment of two on an approved report but see this on the objective edit screen.
> 
> image
> 
> Can you let me know if you can reproduce. I had to manually change DB data to get it to work so it might just be me.

I'm confused as to what you mean. Can you clarify the steps? 
You added both to an AR, deleted one, and then left the AR, went to the edit goals form from the RTR, and that is what you are seeing?

**Review by AdamAdHocTeam — COMMENTED** [review]
Think I found a bug:

1. Create a goal with two objectives (multiple files and topics on each)
2. Create a Approved AR that uses some of the values from each.
3. View the goal on the RTR.

Notice that all values are read only on the first objective even though only some of the existing values are used.

**Comment by AdamAdHocTeam** [comment]
> > Used one objective attachment of two on an approved report but see this on the objective edit screen.
> > image
> > Can you let me know if you can reproduce. I had to manually change DB data to get it to work so it might just be me.
> 
> I'm confused as to what you mean. Can you clarify the steps? You added both to an AR, deleted one, and then left the AR, went to the edit goals form from the RTR, and that is what you are seeing?

As we discussed can disregard this one.

**Comment by thewatermethod** [comment]
> Think I found a bug:
> 
>     1. Create a goal with two objectives (multiple files and topics on each)
> 
>     2. Create a Approved AR that uses some of the values from each.
> 
>     3. View the goal on the RTR.
> 
> 
> Notice that all values are read only on the first objective even though only some of the existing values are used.

Believe this has been nailed down now

**Comment by thewatermethod** [comment]
@AdamAdHocTeam I believe I addressed that issue

**Review by AdamAdHocTeam — COMMENTED** [review]
If I have a goal and objective on a approved AR and move the objective to 'completed'. I no longer have the ability to move the objective back to 'In Progress'.

image

**Review by AdamAdHocTeam — COMMENTED** [review]
This might have been fixed by another PR:

Saving and Editing a goal on a AR clears the anticipated close date.

image

**Comment by thewatermethod** [comment]
@AdamAdHocTeam I believe these have been addressed (feel free to continue your review, but I've also got to add some tests for the updated objective status dropdown)

**Review by AdamAdHocTeam — APPROVED** [review]
This looks good. I ran though the use cases on the mural, finding some minor things along the way that are now fixed.

 :hammer: :fire:

**Review by nvms — APPROVED** [review]
looks and works great!

minor: saw some proptype warnings. I wonder if the proptype for `roles` for these need to be updated as well:

https://github.com/HHS/Head-Start-TTADP/blob/a956ad980b9d6e7880df28ab58c423341bf3450b/frontend/src/components/GoalForm/ReadOnlyGoal.js#L86

https://github.com/HHS/Head-Start-TTADP/blob/a956ad980b9d6e7880df28ab58c423341bf3450b/frontend/src/components/GoalForm/ReadOnlyOtherEntityObjectives.js#L52

like this one:

https://github.com/HHS/Head-Start-TTADP/blob/a956ad980b9d6e7880df28ab58c423341bf3450b/frontend/src/components/GoalForm/ObjectiveForm.js#L221

**Pull request merged**